### PR TITLE
Implemented exercise bank-account

### PR DESCRIPTION
**Pull request by kedeggel**

### CODE REVIEW

File: exercises/bank-account/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    pub fn is_open(&self) -> bool {
+        unimplemented!()
+    }
+
+    pub fn get_balance(&self) -> i32 {
+        unimplemented!()
+    }
+
+    pub fn deposit(&mut self, _: i32) -> Result<i32, BankAccountError> {
```
[the remark this answers is not in the document]
Comment (coriolinus) [reply exercises/bank-account/src/lib.rs]: In the event that the balance ends up being stored as `u64`, we of course don't need the case which tests `i64 -> u64` conversion overflow, though we'd still need the case where the `u64` itself overflows.

**Comment by kedeggel** [comment]
I know this should be a concurrency exercise and I gave thoughts about it. But I couldn't - and still can't - provoke a race condition. 

 >  in order to usefully share a bank account so that multiple threads can manipulate it, its internal state needs to be enclosed in `Cell`s so that many threads can access it simultaneously.

I'm not sure how to implement this. If I enclose balance into a `Cell`, I still don't know how to provoke an error. 
Could you please give me a hint how to implement a unsafe use of a a struct with a "celled" field?
I created a short example, where I

**Comment by coriolinus** [comment]
Here are some attempts to create a compiling example which fails because of race conditions:

- 1st attempt to create a naive, failing test [doesn't compile](https://play.rust-lang.org/?gist=3907d33a38b9549b73c966ed1b5242d6&version=stable) because the `thread_account` can't be mutable
- 2nd attempt to create a naive, failing test [doesn't compile](https://play.rust-lang.org/?gist=911ca4b09bf6ead90c9d5bc11dddd5c7&version=stable) because `Cell` is not `Sync`
- 3rd attempt is no longer naive: we expect this test to pass, because it uses an [`RwLock`](https://doc.rust-lang.org/std/sync/struct.RwLock.html) to guard the internal value. [It passes as expected](https://play.rust-lang.org/?gist=8fac44fa555daad5fa439442923c1a34&version=stable).
- 4th attempt is not naive, but contains a logic error; this version [may or may not pass](https://play.rust-lang.org/?gist=22e530bac5f571d80c7dd293c12287b6&version=stable), depending on whether we get lucky with race conditions. Note that most of the time, this version passes, because I can't use that many threads on the playground machine. 

Safe Rust preventing most race conditions via compile-time errors is a good thing! However, it doesn't (and cannot) prevent all of them. Let's amend my list of expected tests for concurrency:

- One which acts like `test_thread_spam_fails_to_create_race_conditions` by spamming threads and attempting to brute-force a race condition
- One which simply reads the account balance a whole bunch of times simultaneously. Ideally, this has a high enough count that handling it sequentially (i.e. using a `Mutex`) takes perceptibly longer than allowing synchronous access (i.e. using a `RwLock`). 
- More TBD as we think of other ways to attack expected implementations.

**Comment by kedeggel** [comment]
Sorry for not commiting the last weeks, but I've got a lot of work in studies. As soon as I'll have more time I will continue my work on this exercise.

**Comment by coriolinus** [comment]
@kedeggel No problem; we're all volunteers here. As long as you're still interested in implementing this exercise, you can have all the time you need.
